File: Version8/Source/Controls/pyControl.pas
unit pyControl;
{
  ----------------------------------------------------------
  Copyright (c) 2008-2025, Electric Power Research Institute, Inc.
  All rights reserved.
  ----------------------------------------------------------
}
{
  The pyControl is n interface for controllers developped in Python.
  THis structure follows the black box operational principle,
  where it is expected for the Python code to consider some basic
  structure to be compatible with control operations.
}

{$HINTS OFF}
INTERFACE

USES
     Command, ControlClass, ControlElem, CktElement, DSSClass, Arraydef, ucomplex,
     utilities, PointerList, Classes,System.Generics.Collections, Windows, SysUtils;

TYPE

// = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = =
   TpyControl = class(TControlClass)
     private

     protected
        PROCEDURE DefineProperties;
        FUNCTION MakeLike(const pyControlName:String):Integer;Override;
     public
       constructor Create;
       destructor Destroy; override;

       FUNCTION Edit(ActorID : Integer):Integer; override;     // uses global parser
       FUNCTION NewObject(const ObjName:String):Integer; override;

   end;

// = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = =
  TpyControlObj = class(TControlElem)
    private

      pyScript              : String;

    public

      LastCMD      : String;
      constructor Create(ParClass:TDSSClass; const pyControlName:String);
      destructor Destroy; override;

      PROCEDURE MakePosSequence(ActorID : Integer); Override;  // Make a positive Sequence Model
      PROCEDURE RecalcElementData(ActorID : Integer); Override;
      PROCEDURE CalcYPrim(ActorID : Integer); Override;    // Always Zero for a UPFCControl
      PROCEDURE Sample(ActorID : Integer);  Override;    // Sample control quantities and set action times in Control Queue
      PROCEDURE DoPendingAction(Const Code, ProxyHdl:Integer; ActorID : Integer); Override;   // Do the action that is pending from last sample
      PROCEDURE Reset(ActorID : Integer); Override;  // Reset to initial defined state
      PROCEDURE GetCurrents(Curr: pComplexArray; ActorID : Integer); Override; // Get present value of terminal Curr
      PROCEDURE GetInjCurrents(Curr: pComplexArray; ActorID : Integer); Override;   // Returns Injextion currents
      PROCEDURE InitPropertyValues(ArrayOffset:Integer);Override;
      PROCEDURE DumpProperties(Var F:TextFile; Complete:Boolean);Override;
      FUNCTION HandlePIPE(ActorID : Integer): Integer;

  end;


VAR
    ActivepyControlObj:TpyControlObj;

{--------------------------------------------------------------------------}
IMPLEMENTATION

USES

    ParserDel, DSSClassDefs, DSSGlobals, Circuit, uCmatrix, MathUtil, Math;

CONST

    NumPropsThisClass = 1;
    PIPE_FORMAT = '\\%s\pipe\%s'; // \\ServerName\pipe\PipeName
    PIPE_TIMEOUT = 5000;
    BUFF_SIZE = 10000;


{--------------------------------------------------------------------------}
constructor TpyControl.Create;  // Creates superstructure for all UPFCControl objects
Begin
     Inherited Create;

     Class_name   := 'pyControl';
     DSSClassType := DSSClassType + PY_CONTROLLER;

     DefineProperties;

     CommandList := TCommandList.Create(PropertyName, NumProperties);
     CommandList.Abbrev := TRUE;
End;

{--------------------------------------------------------------------------}
destructor TpyControl.Destroy;

Begin
     Inherited Destroy;
End;

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
PROCEDURE TpyControl.DefineProperties;
Begin

     Numproperties := NumPropsThisClass;
     CountProperties;   // Get inherited property count
     AllocatePropertyArrays;


     // Define Property names

     PropertyName^[1] := 'pyScript';


     PropertyHelp^[1] := 'This is the path to the controller script, it is expected to be a python script (*.py).';

     ActiveProperty  := NumPropsThisClass;
     inherited DefineProperties;  // Add defs of inherited properties to bottom of list

End;

{--------------------------------------------------------------------------}
FUNCTION TpyControl.NewObject(const ObjName:String):Integer;
Begin
    // Make a new UPFCControl and add it to UPFCControl class list
    WITH ActiveCircuit[ActiveActor] Do
    Begin
      ActiveCktElement := TpyControlObj.Create(Self, ObjName);
      Result := AddObjectToList(ActiveDSSObject[ActiveActor]);
    End;
End;

{--------------------------------------------------------------------------}
FUNCTION TpyControl.Edit(ActorID : Integer):Integer;
VAR
   ParamPointer : Integer;
   ReDirFile,
   ParamName,
   Param        : String;

Begin

  // continue parsing WITH contents of Parser
  ActivepyControlObj := ElementList.Active;
  ActiveCircuit[ActorID].ActiveCktElement := ActivepyControlObj;

  Result := 0;

  WITH ActivepyControlObj Do Begin

     ParamPointer := 0;
     ParamName := Parser[ActorID].NextParam;
     Param := Parser[ActorID].StrValue;
     WHILE Length(Param)>0 Do Begin
         IF Length(ParamName) = 0 THEN Inc(ParamPointer)
         ELSE ParamPointer := CommandList.GetCommand(ParamName);

         If (ParamPointer>0) and (ParamPointer<=NumProperties)
         THEN PropertyValue[ParamPointer]:= Param;

         CASE ParamPointer OF
            0: DoSimpleMsg('Unknown parameter "' + ParamName + '" for Object "' + Class_Name +'.'+ Name + '"', 364);
            1: Begin
                ReDirFile := ExpandFileName(Param);
                if SysUtils.FileExists(ReDirFile) then
                Begin
                  pyScript := ReDirFile;
                End
                Else
                  DoSimpleMsg('The given path "' + ParamName + '" for Object "' + Class_Name +'.'+ Name + '" does not exist', 36400);

               End
         ELSE
           // Inherited parameters
           ClassEdit( ActivepyControlObj, ParamPointer - NumPropsthisClass)
         End;
         ParamName := Parser[ActorID].NextParam;
         Param := Parser[ActorID].StrValue;
     END;

  End;

End;



{--------------------------------------------------------------------------}
FUNCTION TpyControl.MakeLike(const pyControlName:String):Integer;
VAR
   OtherpyControl:TpyControlObj;
   i:Integer;
Begin
   Result := 0;
   {See if we can find this UPFCControl name in the present collection}
   OtherpyControl := Find(pyControlName);
   IF OtherpyControl<>Nil THEN
   WITH ActivepyControlObj Do Begin

        NPhases := OtherpyControl.Fnphases;
        NConds  := OtherpyControl.Fnconds; // Force Reallocation of terminal stuff

        ElementName       := OtherpyControl.ElementName;
        ControlledElement := OtherpyControl.ControlledElement;  // Pointer to target circuit element
        MonitoredElement  := OtherpyControl.MonitoredElement;  // Pointer to target circuit element

        ElementTerminal   := OtherpyControl.ElementTerminal;


        For i := 1 to ParentClass.NumProperties Do PropertyValue[i] := OtherpyControl.PropertyValue[i];

   End
   ELSE  DoSimpleMsg('Error in pyControl MakeLike: "' + pyControlName + '" Not Found.', 370);

End;




{==========================================================================}
{                    TpyControlObj                                           }
{==========================================================================}
{--------------------------------------------------------------------------}
constructor TpyControlObj.Create(ParClass:TDSSClass; const pyControlName:String);

Begin
  Inherited Create(ParClass);

  Name                := LowerCase(pyControlName);
  DSSObjType          := ParClass.DSSClassType;

End;

destructor TpyControlObj.Destroy;
Begin
     ElementName := '';
     Inherited Destroy;
End;

{--------------------------------------------------------------------------}
PROCEDURE TpyControlObj.RecalcElementData(ActorID : Integer);

VAR
   DevIndex :Integer;

Begin
{Check for existence of monitored element}

  Devindex := GetCktElementIndex(ElementName); // Global function
  IF   DevIndex>0  THEN
  Begin
    MonitoredElement := ActiveCircuit[ActorID].CktElements.Get(DevIndex);
    IF ElementTerminal > MonitoredElement.Nterms THEN
    Begin
      DoErrorMsg('pyControl: "' + Name + '"',
                     'Terminal no. "' +'" does not exist.',
                     'Re-specify terminal no.', 371);
    End
    ELSE
    Begin
     // Sets name of i-th terminal's connected bus in pyControl's buslist
      Setbus(1, MonitoredElement.GetBus(ElementTerminal));
    End;
  End
  ELSE
    DoSimpleMsg('Monitored Element in pyControl.'+Name+ ' does not exist:"'+ElementName+'"', 372);
End;

{--------------------------------------------------------------------------}
procedure TpyControlObj.MakePosSequence(ActorID : Integer);
begin
  if MonitoredElement <> Nil then
  begin
    Nphases := ControlledElement.NPhases;
    Nconds := FNphases;
    Setbus(1, MonitoredElement.GetBus(ElementTerminal));
  end;
  inherited;
end;

{--------------------------------------------------------------------------}
PROCEDURE TpyControlObj.CalcYPrim(ActorID : Integer);
Begin
  // leave YPrims as nil and they will be ignored
  // Yprim is zeroed when created.  Leave it as is.
  //  IF YPrim=nil THEN YPrim := TcMatrix.CreateMatrix(Yorder);
End;

{--------------------------------------------------------------------------}
PROCEDURE TpyControlObj.GetCurrents(Curr: pComplexArray; ActorID : Integer);
VAR
   i:Integer;
Begin

  For i := 1 to Fnconds Do Curr^[i] := CZERO;

End;

PROCEDURE TpyControlObj.GetInjCurrents(Curr: pComplexArray; ActorID : Integer);
Var i:Integer;
Begin
     FOR i := 1 to Fnconds Do Curr^[i] := CZERO;
End;

{--------------------------------------------------------------------------}
PROCEDURE TpyControlObj.DumpProperties(Var F:TextFile; Complete:Boolean);

VAR
   i:Integer;

Begin
  Inherited DumpProperties(F,Complete);

  WITH ParentClass Do
    For i := 1 to NumProperties Do
    Begin
      Writeln(F,'~ ',PropertyName^[i],'=',PropertyValue[i]);
    End;

    If Complete THEN
    Begin
      Writeln(F);
    End;

End;


{--------------------------------------------------------------------------
 Handles the PIPE until the process is done
---------------------------------------------------------------------------}
FUNCTION TpyControlObj.HandlePIPE(ActorID : Integer): Integer;
var
  POnline       : Boolean;
  Written       : Cardinal;
  ClientCmd     : String;

Begin
  POnline := True;
  Result  := 0;
  while (pyServer[ActorID] <> INVALID_HANDLE_VALUE) and (POnline) do
  begin

    ClientCmd := Read_From_PyServer(ActorID);

    if ClientCmd = 'closepipe' then
    Begin
    // This means that the pyscript is done and we need to close the handler
      if LowerCase(LastCMD) = 'yes' then
      Begin
        Result := 1;
      End;
      POnline := False;
    End
    Else
    Begin
      // The py script is sending commands or something different
      LastCMD := ClientCmd;
      if (LowerCase(LastCMD) <> 'yes') and (LowerCase(LastCMD) <> 'no') then
      Begin
        DSSExecutive[ActiveActor].Command :=  ClientCmd;
        //Log('Sending message');
        if GlobalResult = '' then
          GlobalResult := 'OK';

        Write_2_pyServer(GlobalResult, ActorID);
      End
      Else
        Write_2_pyServer('OK', ActorID);

    End;
  end;

End;

{--------------------------------------------------------------------------}
PROCEDURE TpyControlObj.DoPendingAction(Const Code, ProxyHdl:Integer; ActorID : Integer);
begin

  // Do nothing, this is just for coordination with DSS

End;

{--------------------------------------------------------------------------
 Takes a sampe to determine if the control need to perform a control action
---------------------------------------------------------------------------}
PROCEDURE TpyControlObj.Sample(ActorID : Integer);
VAR
  Update      : Boolean;
  LPipeName   : String;
  pHandle     : THandle;

begin

  Update := False; // Default value
 // First, check if the instance's pyServer is running, otherwise do nothing
  if pyServer[ActorID] <> 0 then
  Begin

    Write_2_pyServer(pyScript, ActorID);
    Update := (HandlePIPE(ActorID) = 1);

    {Checks if the controller commands to implement control actions}
    if Update then
    Begin
      // This action is just to sync DSS with the external control action
      With ActiveCircuit[ActorID], ActiveCircuit[ActorID].Solution Do
        ControlQueue.Push(DynaVars.intHour, DynaVars.t, 0, 0, Self, ActorID);

    End;
    GlobalResult := '';
  End;
End;

procedure TpyControlObj.InitPropertyValues(ArrayOffset: Integer);
begin

  PropertyValue[1]  := '""';   //'pyScript';
  inherited  InitPropertyValues(NumPropsThisClass);

end;

procedure TpyControlObj.Reset;
begin
  // inherited;

end;

end.

File: trunk/VersionC/Controls/pyControl.cpp

#pragma hdrstop

#include "pyControl.h"

#include "ParserDel.h"
#include "DSSClassDefs.h"
#include "DSSGlobals.h"
#include "Circuit.h"
#include "Ucmatrix.h"
#include "mathutil.h"
#include <math.h>
#include "Utilities.h"

using namespace std;
using namespace Arraydef;
using namespace Circuit;
using namespace CktElement;
using namespace Command;
using namespace ControlClass;
using namespace ControlElem;
using namespace DSSClass;
using namespace DSSClassDefs;
using namespace DSSGlobals;
using namespace DSSObject;
using namespace ParserDel;
using namespace System;
using namespace UPFC;
using namespace Ucmatrix;
using namespace Ucomplex;
using namespace mathutil;
using namespace Utilities;

namespace pyControl
{

TpyControlObj::TpyControlObj(DSSClass::TDSSClass* ParClass) : inherited(ParClass) {}
TpyControlObj::TpyControlObj(String ClassName) : inherited(ClassName) {}
TpyControlObj::TpyControlObj() {}


TpyControlObj* ActivepyControlObj = nullptr;
const int NumPropsThisClass = 1;
const string PIPE_FORMAT = "\\\\% s\\pipe\\% s"; // \\ServerName\pipe\PipeName 
const int PIPE_TIMEOUT = 5000;
const int BUFF_SIZE = 10000;


/*--------------------------------------------------------------------------*/  // Creates superstructure for all UPFCControl objects

TpyControl::TpyControl()
{
	;
	Class_Name = "pyControl";
	DSSClassType = DSSClassType + PY_CONTROLLER;
	DefineProperties();
	auto&& slc = Slice(PropertyName, NumProperties);
	CommandList = TCommandList(slc.data(), NumProperties);
	CommandList.set_AbbrevAllowed(true);
}

/*--------------------------------------------------------------------------*/

TpyControl::~TpyControl()
{
	// inherited::Destroy();
}


//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

void TpyControl::DefineProperties()
{
	NumProperties = NumPropsThisClass;
	CountProperties();   // Get inherited property count
	AllocatePropertyArrays();


     // Define Property names
	PropertyName[1 - 1] = "pyScript";
	PropertyHelp[1 - 1] = "This is the path to the controller script, it is expected to be a python script (*.py). ";

	ActiveProperty = NumPropsThisClass - 1;
	inherited::DefineProperties();  // Add defs of inherited properties to bottom of list
}

/*--------------------------------------------------------------------------*/

int TpyControl::NewObject(const String ObjName)
{
	int result = 0;
    // Make a new pyControl and add it to pyControl class list
	/*# with ActiveCircuit[ActiveActor] do */
	{
		
		ActiveCircuit[ActiveActor]->Set_ActiveCktElement(new TpyControlObj(this, ObjName));
		result = AddObjectToList(ActiveDSSObject[ActiveActor]);
	}
	return result;
}

/*--------------------------------------------------------------------------*/

int TpyControl::Edit(int ActorID)
{
	int result = 0;
	int ParamPointer = 0;
	String ParamName;
	String Param;
	int i = 0;

  // continue parsing WITH contents of Parser
	ActivepyControlObj = (TpyControlObj*) ElementList.Get_Active();
	ActiveCircuit[ActorID]->Set_ActiveCktElement(ActivepyControlObj);
	result = 0;
	/*# with ActivepyControlObj do */
	{
		auto with0 = ActivepyControlObj;
		ParamPointer = 0;
		ParamName = Parser[ActorID]->GetNextParam();
		Param = Parser[ActorID]->MakeString_();
		while(Param.size() > 0)
		{
			if(ParamName.size() == 0)
				++ParamPointer;
			else
				ParamPointer = CommandList.Getcommand(ParamName);
			if((ParamPointer > 0) && (ParamPointer <= NumProperties))
				with0->Set_PropertyValue(ParamPointer,Param);
			switch(ParamPointer)
			{
				case 	0:
				DoSimpleMsg(String("Unknown parameter \"") + ParamName
	           + "\" for Object \""
	           + Class_Name
	           + "."
	           + with0->get_Name()
	           + "\"", 364);
				break;
				case 1:
				{
					String RedirFile = ExpandFileName(Param);
					if (FileExists(RedirFile))
						with0->pyScript = RedirFile; 
					else
                        DoSimpleMsg("The given path '" + ParamName + "' for Object '" + Class_Name + "." + with0->get_Name() + "' does not exist", 36400);
				}
				break;
           // Inherited parameters
				default:
				ClassEdit(ActiveUPFCControlObj, ParamPointer - NumPropsThisClass);
				break;
			}
            ParamName = Parser[ActorID]->GetNextParam();
            Param = Parser[ActorID]->MakeString_();
		}
	}

  //RecalcElementData(ActorID);
	return result;
}



/*--------------------------------------------------------------------------*/

int TpyControl::MakeLike(const String pyControlName)
{
	int result = 0;
	TpyControlObj* OtherpyControl = nullptr;
	int i = 0;
	result = 0;
   /*See if we can find this UPFCControl name in the present collection*/
	OtherpyControl = ((TpyControlObj*) Find(pyControlName));
        if (OtherpyControl != nullptr)
		/*# with ActiveUPFCControlObj do */
		{
			auto with0 = ActiveUPFCControlObj;
			int stop = 0;
			with0->Set_NPhases(OtherpyControl->Fnphases);
            with0->Set_Nconds(OtherpyControl->Fnconds); // Force Reallocation of terminal stuff
            with0->ElementName = OtherpyControl->ElementName;
            with0->Set_ControlledElement(OtherpyControl->get_FControlledElement()); // Pointer to target circuit element
            with0->Set_MonitoredElement(OtherpyControl->get_FMonitoredElement()); // Pointer to target circuit element
            with0->ElementTerminal = OtherpyControl->ElementTerminal;
			for(stop = with0->ParentClass->NumProperties, i = 1; i <= stop; i++)
			{
				with0->Set_PropertyValue(i,OtherpyControl->Get_PropertyValue(i));
			}
		}
	else
		DoSimpleMsg(String("Error in UPFCControl MakeLike: \"") + pyControlName
	           + "\" Not Found.", 370);
	return result;
}




/*==========================================================================*/
/*                    TpyControlObj                                           */
/*==========================================================================*/
/*--------------------------------------------------------------------------*/

TpyControlObj::TpyControlObj(TDSSClass* ParClass, const String pyControlName)
 : inherited(ParClass)
{
	Set_Name(LowerCase(pyControlName));
	DSSObjType = ParClass->DSSClassType;
    LastCMD = "";
    pyScript = "";
}

TpyControlObj::~TpyControlObj()
{
	ElementName = "";
	// inherited::Destroy();
}

/*--------------------------------------------------------------------------*/

void TpyControlObj::RecalcElementData(int ActorID)
{

/*Do nothing*/

}

/*--------------------------------------------------------------------------*/

void TpyControlObj::MakePosSequence(int ActorID)
{
	// This is here but it does nothing
	if(get_FMonitoredElement() != nullptr)
	{
		Set_NPhases(get_FControlledElement()->Get_NPhases());
		Set_Nconds(Fnphases);
		SetBus(1, get_FMonitoredElement()->GetBus(ElementTerminal));
	}
	inherited::MakePosSequence(ActorID);
}

/*--------------------------------------------------------------------------*/

void TpyControlObj::CalcYPrim(int ActorID)
{

  // leave YPrims as nil and they will be ignored
  // Yprim is zeroed when created.  Leave it as is.
  //  IF YPrim=nil THEN YPrim := TcMatrix.CreateMatrix(Yorder);
}

/*--------------------------------------------------------------------------*/

void TpyControlObj::GetCurrents(pComplexArray Curr, int ActorID)
{
	int i = 0;
	int stop = 0;
	for(stop = Fnconds, i = 1; i <= stop; i++)
	{
		(Curr)[i - 1] = CZero;
	}
}

void TpyControlObj::GetInjCurrents(pComplexArray Curr, int ActorID)
{
	int i = 0;
	int stop = 0;
	for(stop = Fnconds, i = 1; i <= stop; i++)
	{
		(Curr)[i - 1] = CZero;
	}
}

/*--------------------------------------------------------------------------*/

void TpyControlObj::DumpProperties(TTextRec& f, bool Complete)
{
	int i = 0;
	inherited::DumpProperties(f, Complete);
	/*# with ParentClass do */
	{
		auto with0 = ParentClass;
		for (i = 1; i <= with0->NumProperties; i++)
		{
			{ Write(f, "~ "); Write(f, with0->PropertyName[i - 1]); Write(f, "="); WriteLn(f, Get_PropertyValue(i)); }
		}
	}
	if(Complete)
	{
		WriteLn(f);
	}
}

/*--------------------------------------------------------------------------*/
int TpyControlObj::HandlePIPE(int ActorID)
{
    bool POnline = true;
    int Written = 0;
    int Result = 0;
    String ClientCmd = "";
#ifdef windows
	//Windows routine
	while ((pyServer[ActorID] != INVALID_HANDLE_VALUE) && POnline)
    {
		ClientCmd = Read_From_PyServer(ActorID);
        if (ClientCmd == "closepipe")
        {
            // This means that the pyscript is done and we need to close the handler
            if (LowerCase(LastCMD) == "yes")
            {
                Result = 1;
            }
            POnline = false;
        }
        else
        {
            // The py script is sending commands or something different
            LastCMD = ClientCmd;
            if ((LowerCase(LastCMD) != "yes") && (LowerCase(LastCMD) != "no"))
            {
                DSSExecutive[ActorID]->Set_Command(ClientCmd);
                if (GlobalResult.empty())
                    GlobalResult = "OK";

                Write_2_PyServer(GlobalResult, ActorID);
            }
			else
                Write_2_PyServer("OK", ActorID);
        }
    }
#elif __linux__
    // Windows routine
    while ((pyServer[ActorID] != "") && POnline)
    {
        ClientCmd = Read_From_PyServer(ActorID);
        if (ClientCmd == "closepipe")
        {
            // This means that the pyscript is done and we need to close the handler
            if (LowerCase(LastCMD) == "yes")
            {
                Result = 1;
            }
            POnline = false;
        }
        else
        {
            // The py script is sending commands or something different
            LastCMD = ClientCmd;
            if ((LowerCase(LastCMD) != "yes") && (LowerCase(LastCMD) != "no"))
            {
                DSSExecutive[ActorID]->Set_Command(ClientCmd);
                if (GlobalResult.empty())
                    GlobalResult = "OK";

                Write_2_PyServer(GlobalResult, ActorID);
            }
            else
                Write_2_PyServer("OK", ActorID);
        }
    }
#endif
    return Result;
}

/*--------------------------------------------------------------------------*/

void TpyControlObj::DoPendingAction(int Code, int ProxyHdl, int ActorID)
{
	// Do nothing
}

/*--------------------------------------------------------------------------*/

void TpyControlObj::sample(int ActorID)
{
	bool Update = false;
    String LPipeName = "";

#ifdef windows
    HANDLE pHandle = NULL;

	// First, check if the instance's pyServer is running, otherwise do nothing

    if (ASSIGNED(pyServer[ActorID]))
    {
        Write_2_PyServer(pyScript, ActorID);
        Update = (HandlePIPE(ActorID) == 1);

        /*Checks if the controller commands to implement control actions*/
        if (Update)
        {
            // This action is just to sync DSS with the external control action
            auto with0 = ActiveCircuit[ActorID];
            auto with1 = with0->Solution;

            with0->ControlQueue.Push(with1->DynaVars.intHour, with1->DynaVars.T, 0, 0, this, ActorID);
        }
        GlobalResult = "";
    }
#elif __linux__
    if (!(pyServer[ActorID].empty()))
    {
        Write_2_PyServer(pyScript, ActorID);
		Update = (HandlePIPE(ActorID) == 1);

		/*Checks if the controller commands to implement control actions*/
		if (Update)
		{
			// This action is just to sync DSS with the external control action
            auto with0 = ActiveCircuit[ActorID];
			auto with1 = with0->Solution;

			with0->ControlQueue.Push(with1->DynaVars.intHour, with1->DynaVars.T, 0, 0, this, ActorID);
		}
		GlobalResult = "";
    }
#endif
}

void TpyControlObj::InitPropertyValues(int ArrayOffset)
{
	Set_PropertyValue(1,"");   //'pyScript';
	inherited::InitPropertyValues(NumPropsThisClass);
}

void TpyControlObj::Reset(int ActorID)
{

  // inherited;
}




}  // namespace UPFCControl





